node/src/components/small_network/task_manager.rs

```diff
@@ -0,0 +1,231 @@
+            // join handles map, if still in there. If the lock is poisoned we skip this step.
+            if let Ok(mut join_handles) = join_handles.lock() {
+                join_handles.remove(&task_id);
+            }
+        });
+
+        // Now that the task has been spawned, insert into join handle list.
+        if let Ok(mut join_handles) = self.join_handles.lock() {
+            join_handles.insert(task_id, join_handle);
+        } else {
+            // This is problematic, but no reason to crash in some cases.
+            warn!("join handles lock has been poisoned, no freeing entry");
+        }
+    }
+
+    /// Waits for all background tasks to shut down.
+    ///
+    /// Returns the number of tasks that were not joined successfully in time.
+    ///
+    /// If the lock has been poisoned, `Err(())` is returned, as it is impossible to properly join
+    /// the tasks at that point. They will still be asked to shutdown.
+    pub async fn shutdown_and_wait(self, timeout: Duration) -> Result<usize, ()> {
+        // Dropping the sender will cause all receivers to shutdown.
+        drop(self.shutdown_sender);
+
+        // We create a stream of join results that we can collect in parallel.
+        let mut joins: FuturesUnordered<_> = {
+            let mut guard = self.join_handles.lock().map_err(drop)?;
+            // Note: It is important to drop the lock as soon as possible, as the dropped shutdown
+            // sender above will *likely* not have any effects until the next `.await`. Otherwise
+            // tasks that want to update after exiting will not be able to acquire the lock and time
+            // out waiting to do so.
+            guard
+                .drain()
+                .map(|(task_id, join_handle)| async move {
+                    (
+                        task_id,
+                        match time::timeout(timeout, join_handle).await {
+                            Ok(Ok(())) => ShutdownOutcome::Ok,
+                            Ok(Err(join_err)) => ShutdownOutcome::JoinFailure(join_err),
+                            Err(_) => ShutdownOutcome::TimedOut,
+                        },
+                    )
+                })
+                .collect()
+        };
+
+        // Collect all the join handles and output appropriate errors.
+        let mut failed = 0;
+        while let Some((task_id, outcome)) = joins.next().await {
+            match outcome {
+                ShutdownOutcome::Ok => {
+                    trace!(task_id, "successfully joined task");
+                }
+                ShutdownOutcome::JoinFailure(err) => {
+                    warn!(task_id, %err, "could not join task");
+                    failed += 1;
+                }
+                ShutdownOutcome::TimedOut => {
+                    error!(task_id, "task joining timed out");
+                    failed += 1;
+                }
+            }
+        }
```

Comment:
Couldn't we just `.map` here and then `.sum`?

Reply:
I actually prefer this version. It's more readable IMO and makes it more obvious what we're doing.

Reply:
> Couldn't we just .map here and then .sum?

It's a stream, not an iterator, so out-of-the-box, I don't think there's a `.sum` impl available? The only one I am aware of is https://docs.rs/async-std/1.9.0/async_std/stream/trait.Sum.html.

If it was available, imagine the code'd look like this

```rust
        Ok(joins
            .map(|(task_id, outcome)| match outcome {
                ShutdownOutcome::Ok => {
                    trace!(task_id, "successfully joined task");
                    0
                }
                ShutdownOutcome::JoinFailure(err) => {
                    warn!(task_id, %err, "could not join task");
                    1
                }
                ShutdownOutcome::TimedOut => {
                    error!(task_id, "task joining timed out");
                    1
                }
            })
            .sum());
```

Overall, I don't think the improvement is that great, I can certainly see both. The explicit version is a little easier to maintain/debug if things change in my mind; the pattern of mapping-for-side-effects has always felt weird to me.

Reply:
Side note: This code used to run `.for_each_concurrent`, but with the change to `FuturesUnordered` as recommend by the docs, I simplified it a bit.